librad/src/meta/project.rs

```diff
@@ -87,6 +154,153 @@ impl Project {
         xs.dedup();
         self.maintainers = NonEmpty::from_slice(&xs).unwrap();
     }
+
+    pub fn canonical_data(&self) -> Result<Vec<u8>, Error> {
+        let mut cleaned = self.clone();
+        cleaned.signatures.clear();
+        let mut buffer: Vec<u8> = vec![];
+        let mut ser =
+            serde_json::Serializer::with_formatter(&mut buffer, CanonicalFormatter::new());
+        cleaned
+            .serialize(&mut ser)
+            .map_err(Error::SerializationFailed)?;
+        Ok(buffer)
+    }
+
+    pub fn canonical_text_contents(&self) -> Result<String, Error> {
+        String::from_utf8(self.canonical_data()?).map_err(|_| Error::InvalidUtf8)
+    }
+
+    pub fn sign(&self, key: &Key) -> Result<Signature, Error> {
+        Ok(key.sign(&self.canonical_data()?))
+    }
+
+    pub fn build_signature(&self, key: &Key) -> Result<ProjectSignature, Error> {
+        let signature = self.sign(key)?;
+        Ok(ProjectSignature {
+            key: PeerId::from(key.clone()),
+            sig: signature.encode_hex_upper(),
+        })
+    }
+
+    pub fn add_signature(&mut self, key: &Key) -> Result<(), Error> {
+        let id = PeerId::from(key.clone());
+        if self.signatures.iter().any(|s| s.key == id) {
+            Err(Error::SignatureAlreadyPresent)
+        } else if !self.maintainers.iter().any(|m| m == &id) {
+            Err(Error::SignatureFromNonMaintainer)
+        } else {
+            self.signatures.push(self.build_signature(key)?);
+            Ok(())
+        }
+    }
+
+    pub fn remove_signature(&mut self, id: &PeerId) {
+        self.signatures.retain(|s| s.key != *id);
+    }
+
+    pub fn verify_signature(&self, signature: &ProjectSignature) -> Result<(), Error> {
+        signature.verify_data(&self.canonical_data()?)
+    }
+
+    pub fn verify_signatures(&self) -> Result<(), Error> {
+        let data = self.canonical_data()?;
+        for s in self.signatures.iter() {
+            s.verify_data(&data)?
+        }
+        Ok(())
+    }
+
+    pub fn check_signatures_against_maintainers(&self) -> Result<(), Error> {
+        let mut maintainers = HashSet::<&PeerId>::new();
+        for m in self.maintainers.iter() {
+            maintainers.insert(m);
+        }
+
+        for s in self.signatures.iter() {
+            let m = maintainers.take(&s.key);
+            match m {
+                Some(_) => {}
+                None => return Err(Error::SignatureFromNonMaintainer),
+            }
+        }
+
+        if !maintainers.is_empty() {
+            return Err(Error::SignatureMissing);
+        }
+
+        Ok(())
+    }
+
+    pub fn check_validity(&self) -> Result<(), Error> {
+        self.check_signatures_against_maintainers()?;
+        self.verify_signatures()?;
```

Comment:
True.
This is just hair splitting and I can do it either way.

However in this case I personally like it more with the explicit `Ok(())` at the end because all the checks are expressed exactly in the same way in a sequence of invocations ending with the `?` operator:
```
check_something()?
check_something_else()?
// more checks could follow...
```
This way when reading the last check is not a "special case".

Reply:
Good point, that makes perfect sense 👌